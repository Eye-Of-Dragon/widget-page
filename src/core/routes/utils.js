// @flow
import * as React from "react";
import { Route } from "react-router-dom";

const ACCESS = {
  PRIVATE: 'private',
  PUBLIC:'public'
};

export type RoutesProps = {
  component: React.Element;
  path: string;
  access:  ACCESS.PRIVATE | ACCESS.PUBLIC;
  exact: boolean;
  roles?: string[];
}

export type Props = {
  component: React.Node;
  path: string;
  store: any;
};

export const PrivateRoute = (props: Props) => {
  const { component: Component, store, ...rest } = props;
  return (
    <Route
      {...rest}
      render={props => <Component {...props} />}
    />

  );
};


export const MapRoutes = (routes: RoutesProps[], Layout: any = React.Fragment) =>
  routes
    .filter(/*apply role*/_ => _)
    .map(props => {
      const { component, ...rest } = props;
      const { access } = rest;
      if (access === ACCESS.PRIVATE) {
        return ({ component: props => Layout({ Component: () => PrivateRoute({ component, ...props })} ),  ...rest });
      }
      return props;
    });
