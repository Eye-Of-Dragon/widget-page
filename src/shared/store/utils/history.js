import { createBrowserHistory } from 'history/index';
const history = createBrowserHistory();
export default history;
