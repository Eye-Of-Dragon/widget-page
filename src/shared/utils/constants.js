export const PATHS = {
  widgets: "/widgets",
  widgetAdd: "/widget/add",
  home: "/",
};
