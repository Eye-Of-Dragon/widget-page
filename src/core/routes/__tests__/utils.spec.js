import React from 'react';
import { render } from '@testing-library/react';
import { MapRoutes, PrivateRoute } from '../utils';
import type { RoutesProps } from "../utils";
import Dropdown from "../../../shared/components/DropDown";

jest.mock(
  "react-router-dom",
  (): any => {
    const Route = ({ render: RenderComponents, ...rest }) => (
        <div {...rest}>
          <RenderComponents />
        </div>
      );

    return {
      Route
    };
  },
);

describe('PrivateRoute && MapRoutes', () => {
  test('PrivateRoute snap-shot ', () => {
    const { container } = render(<PrivateRoute
      component={() => <div>component</div>}
      store={undefined}
      path={'/home'}
    />);
    expect(container).toMatchSnapshot();
  });

  test('MapRoutes snap-shot ', () => {
    const routes: RoutesProps[] = [
      {
        component: () => <div>home</div>,
        path: '/home',
        exact: true,
        access: 'public',
      },
      {
        component: () => <div>products</div>,
        path: '/products',
        exact: true,
        access: 'private',
      }
    ];

    const ProtectedLayout = ({ Component, ...props }) => <div role={'private'}><Component {...props}/></div>;
    const mapedRoutes: any[] = MapRoutes(routes, ProtectedLayout);
    const RenderedComponent = () => (
      <div>
        {mapedRoutes.map(({ component: Component }, index) => (<Component key={index}/>))}
      </div>
    );
    const { container } = render(<RenderedComponent />);
    expect(mapedRoutes.length).toBe(2);
    expect(container).toMatchSnapshot();
  });
});
