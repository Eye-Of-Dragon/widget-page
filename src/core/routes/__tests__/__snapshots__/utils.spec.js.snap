// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrivateRoute && MapRoutes MapRoutes snap-shot  1`] = `
<div>
  <div>
    <div>
      home
    </div>
    <div
      role="private"
    >
      <div>
        <div>
          products
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`PrivateRoute && MapRoutes PrivateRoute snap-shot  1`] = `
<div>
  <div
    path="/home"
  >
    <div>
      component
    </div>
  </div>
</div>
`;
